### PR TITLE
Init vx-delegate

### PR DESCRIPTION
**Pull request by liyuenan2333**
Signed-off-by: xiang.zhang <xiang.zhang@verisilicon.com>

### CODE REVIEW

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+    compiled_ = graph_->Compile();
+    if (!compiled_) {
+      LOG(FATAL) << "Failed to verify graph";
+      return kTfLiteDelegateError;
+    }
+
+    LOG(INFO) << "Verified graph";
+  }
+
+  // TODO(derekjchow): Return error if compilation failed.
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:508]: remove this? already fatal in line 501

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+  for (int tensor_idx : op_data.subgraph_inputs) {
+    const TfLiteTensor& tf_tensor = context->tensors[tensor_idx];
+    LOG(INFO) << "Copying input " << tensor_idx << ":"
+              << tf_tensor.name;
+    auto* tensor = tensors_[tensor_idx].get();
+    if (!tensor) {
+      LOG(FATAL) << "Disaster!";
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:515]: more descriptive log

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+    }
+
+    const void* tensor_data =
+        reinterpret_cast<const void*>(tf_tensor.data.raw_const);
+    // TODO(derekjchow): Check result
+    tensor->CopyDataToTensor(const_cast<void*>(tensor_data));
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:521]: check return value

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+  }
+
+  LOG(INFO) << "Invoking graph";
+  if (!graph_->Run()) {
+    LOG(FATAL) << "Failed to verify graph";
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:526]: wrong message. should be "failed to run graph"

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+  }
+
+  for (int tensor_idx : op_data.subgraph_outputs) {
+    TfLiteTensor& tf_tensor = context->tensors[tensor_idx];
+    LOG(INFO) << "Copying output " << tensor_idx << ":"
+              << tf_tensor.name;
+    auto* tensor = tensors_[tensor_idx].get();
+    if (!tensor) {
+      LOG(FATAL) << "Disaster!";
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:535]: descriptive message

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+    }
+
+    void* tensor_data = reinterpret_cast<void*>(tf_tensor.data.raw);
+    // TODO(derekjchow): Check result
+    tensor->CopyDataFromTensor(tensor_data);
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:540]: check return value

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+  }
+
+  // Copy output states to input states
+  for (int tensor_idx : op_data.subgraph_states) {
+    TfLiteTensor& tf_tensor = context->tensors[tensor_idx];
+    LOG(INFO) << "Copying state " << tensor_idx << ":"
+              << tf_tensor.name;
+    auto* tensor = state_tensors_[tensor_idx].get();
+    if (!tensor) {
+      LOG(FATAL) << "Disaster!";
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:550]: descriptive error message

File: tensorflow/lite/delegates/vx-delegate/delegate_main.cc [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+    }
+
+    void* tensor_data = reinterpret_cast<void*>(tf_tensor.data.raw);
+    tensor->CopyDataFromTensor(tensor_data);
+  }
+
+  // Mark all the data buffers that need copying back.
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.cc:557]: remove unused code

File: tensorflow/lite/delegates/vx-delegate/delegate_main.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
+#ifndef OVXLIBXX_DELEGATE_H_
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.h:1]: guard : TENSORFLOW_LITE_DELEGATES_VX_DELEGAGE_DELEGATE_MAIN_H

File: tensorflow/lite/delegates/vx-delegate/delegate_main.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+#include "tensorflow/lite/builtin_op_data.h"
+#include "tensorflow/lite/builtin_ops.h"
+#include "tensorflow/lite/context.h"
+
+#include "tim/vx/context.h"
+#include "tim/vx/graph.h"
+#include "tim/vx/operation.h"
+#include "tim/vx/tensor.h"
+
+namespace ovxlibxx_delegate {
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.h:18]: reorg namespace

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 11]
```diff
@@ -0,0 +1,1459 @@
...
+
+  return tim::vx::PadType::AUTO;
+}
+
+/// Insert activation layer before the `original_output_tensor`
+/// Return the input tensor of new activation layer
+std::shared_ptr<tim::vx::Tensor> ProcessFusedActivation(
+    ovxlibxx_delegate::Delegate* delegate,
+    TfLiteFusedActivation fused_activation,
+    std::shared_ptr<tim::vx::Tensor> original_output_tensor) {
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:69]: const .. &

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 12]
```diff
@@ -0,0 +1,1459 @@
...
+  std::shared_ptr<tim::vx::Operation> op = nullptr;
+  switch (fused_activation) {
+    default:
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:72]: move default to the end?

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 13]
```diff
@@ -0,0 +1,1459 @@
...
+  delegate->GetTensors().push_back(temp);
+
+  return temp;
+}
+
+// Insert a permute node after the `original_input_tensor`
+std::shared_ptr<tim::vx::Tensor> PermuteInputTensor(
+    ovxlibxx_delegate::Delegate* delegate,
+    std::shared_ptr<tim::vx::Tensor> original_input_tensor,
+    std::vector<uint32_t> perm) {
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:109]: const .. &

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 14]
```diff
@@ -0,0 +1,1459 @@
...
+  delegate->GetOps().push_back(op);
+  delegate->GetTensors().push_back(temp);
+
+  return temp;
+}
+
+// Insert a permute node after the `original_input_tensor`
+std::shared_ptr<tim::vx::Tensor> PermuteInputTensor(
+    ovxlibxx_delegate::Delegate* delegate,
+    std::shared_ptr<tim::vx::Tensor> original_input_tensor,
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:108]: const ref

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 15]
```diff
@@ -0,0 +1,1459 @@
...
+      break;
+    case kTfLiteActTanh:
+      op = delegate->GetGraph()->CreateOperation<tim::vx::ops::Tanh>();
+      break;
+    case kTfLiteActSigmoid:
+      op = delegate->GetGraph()->CreateOperation<tim::vx::ops::Sigmoid>();
+      break;
+  }
+
+  auto temp = delegate->GetGraph()->CreateTensor(
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:93]: don't name it as temp

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 16]
```diff
@@ -0,0 +1,1459 @@
...
+  delegate->GetOps().push_back(op);
+  delegate->GetTensors().push_back(temp);
+
+  return temp;
+}
+
+// Insert a permute node before the `original_output_tensor`
+std::shared_ptr<tim::vx::Tensor> PermuteOutputTensor(
+    ovxlibxx_delegate::Delegate* delegate,
+    std::shared_ptr<tim::vx::Tensor> original_output_tensor,
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:134]: const ref

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 17]
```diff
@@ -0,0 +1,1459 @@
...
+        return false;
+      }
+      for (int j = 0; j < context->tensors[output_index].dims->size; j++) {
+        if (context->tensors[output_index].dims->data[j] == 0) {
+          LOG(INFO) << "ovxlibxx doesn't support the tensor has 0 dim";
+          return false;
+        }
+      }
+    }
+
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:297]: only single empty line is allowed

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 18]
```diff
@@ -0,0 +1,1459 @@
...
+      TfLiteNode* node, const TfLiteRegistration* registration) const {
+    return false;
+  }
+};
+
+template<typename T_OperationType>
+struct ReduceOpMapper : public OpMapperBase<TfLiteReducerParams> {
+  std::string name_;
+
+  ReduceOpMapper(std::string name) : name_(name) {}
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:1257]: const ref

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 19]
```diff
@@ -0,0 +1,1459 @@
...
+    }
+
+    status = HandleMapOp(delegate, inputs, outputs, states, params);
+
+    return status;
+  }
+
+  virtual bool HandleMapOp(
+      ovxlibxx_delegate::Delegate* delegate,
+      std::vector<std::shared_ptr<tim::vx::Tensor>>& inputs,
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:335]: const ?

File: tensorflow/lite/delegates/vx-delegate/op_map.h [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+#include <map>
+#include <memory>
+#include <vector>
+#include <functional>
+
+#include "delegate_main.h"
+#include "tim/vx/operation.h"
+
+namespace ovxlibxx_op_map {
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.h:12]: vx::delegate::mapping

File: tensorflow/lite/delegates/vx-delegate/util.h [hunk 21]
```diff
@@ -0,0 +1,65 @@
+#ifndef OVXLIBXX_UTIL_H_
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/util.h:1]: GUARD name

File: tensorflow/lite/delegates/vx-delegate/util.h [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+#define OVXLIBXX_UTIL_H_
+
+#include <stdint.h>
+#include <vector>
+
+#include "tensorflow/core/platform/logging.h"
+
+namespace ovxlibxx_util {
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/util.h:9]: vx::delegate::utils

File: tensorflow/lite/delegates/vx-delegate/util.h [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+  for (int i = 0; i < perm.size(); i++) {
+    output[i] = input[perm[i]];
+  }
+
+  return output;
+}
+
+}  // namespace ovxlibxx_util
+
+#endif
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/util.h:65]: new line

File: tensorflow/lite/kernels/test_util.cc [hunk 24]
```diff
@@ -223,6 +230,14 @@ TfLiteStatus SingleOpModel::ApplyDelegate() {
     delegate_ = TestNnApiDelegate();
   }
 
+  if (force_use_vx_delegate) {
+    delegate_ = ::ovxlibxx_delegate::Delegate::Create();
+  }
+
+  if (force_use_nnapi && force_use_vx_delegate) {
+    // TODO warning or error
```
Comment (sunshinemyson) [inline tensorflow/lite/kernels/test_util.cc:238]: log(FATAL) << "Don't setup ..."

File: tensorflow/lite/kernels/test_util.cc [hunk 25]
```diff
@@ -249,6 +264,17 @@ void SingleOpModel::SetForceUseNnapi(bool use_nnapi) {
 // static
 bool SingleOpModel::GetForceUseNnapi() { return force_use_nnapi; }
 
+// static
+void SingleOpModel::SetForceUseVxDelegate(bool use_vsi) {
```
Comment (sunshinemyson) [inline tensorflow/lite/kernels/test_util.cc:268]: rename use_vsi to use_vx_delegate

File: tensorflow/lite/kernels/test_util.h [hunk 26]
```diff
@@ -516,6 +516,10 @@ class SingleOpModel {
   // Enables NNAPI delegate application during interpreter creation.
   static void SetForceUseNnapi(bool use_nnapi);
   static bool GetForceUseNnapi();
+
+  static void SetForceUseVxDelegate(bool use_vsi);
```
Comment (sunshinemyson) [inline tensorflow/lite/kernels/test_util.h:520]: rename use_vsi to use_vx_deleagate

**Review by sunshinemyson — CHANGES_REQUESTED** [review]
1. license missed
2. const ref missed
3. reorg namespace

File: tensorflow/lite/delegates/vx-delegate/util.h (hunk 23 above)
Comment (liyuenan2333) [reply tensorflow/lite/delegates/vx-delegate/util.h:65]: what it means?

File: tensorflow/lite/kernels/test_util.cc (hunk 24 above)
Comment (liyuenan2333) [reply tensorflow/lite/kernels/test_util.cc:238]: what it mean?

File: tensorflow/lite/delegates/vx-delegate/delegate_main.h [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+  std::vector<std::shared_ptr<tim::vx::Tensor>>& GetTensors() {
+    return tensors_;
+  }
+
+ private:
+  struct OperationDataType {
+    int builtin_code;
+    std::vector<int> inputs;
+    std::vector<int> outputs;
+    std::vector<int> states;
```
[the remark this answers is not in the document]
Comment (sunshinemyson) [reply tensorflow/lite/delegates/vx-delegate/delegate_main.h:75]: //!< for recurrent operation: LSTM, RNN

File: tensorflow/lite/delegates/vx-delegate/delegate_main.h [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+  std::unique_ptr<Delegate> delegate;
+};
+
+class Delegate {
+ public:
+  static TfLiteDelegate* Create();
+  static bool SupportedOp(TfLiteContext* context,
+      TfLiteNode* node, const TfLiteRegistration* registration);
+
+  Delegate();
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/delegate_main.h:51]: make it private

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 29]
```diff
@@ -0,0 +1,1472 @@
...
+
+template<typename T_OperationType>
+struct ReduceOpMapper : public OpMapperBase<TfLiteReducerParams> {
+  std::string name_;
+
+  ReduceOpMapper(std::string name) : name_(name) {}
+  virtual bool HandleMapOp(
+      vx::delegate::Delegate* delegate,
+      std::vector<std::shared_ptr<tim::vx::Tensor>>& inputs,
+      std::vector<std::shared_ptr<tim::vx::Tensor>>& outputs,
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:1273]: const ref

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 30]
```diff
@@ -0,0 +1,1472 @@
...
+    auto op = delegate->GetGraph()->CreateOperation<T_OperationType>();
+    (*op).BindInputs(inputs).BindOutputs(outputs);
+
+    delegate->GetOps().push_back(std::move(op));
+
+    return true;
+  }
+};
+
+typedef std::function<std::unique_ptr<IOpMapper>()> createIOpMapItemFunc;
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:1340]: using

File: tensorflow/lite/delegates/vx-delegate/op_map.cc [hunk 31]
```diff
@@ -0,0 +1,1472 @@
...
+  REGISTER_OP_MAPPER(kTfLiteBuiltinReduceProd, \
+      ReduceOpMapper<tim::vx::ops::ReduceProd>, "Prod"),
+  REGISTER_OP_MAPPER(kTfLiteBuiltinMean, \
+      ReduceOpMapper<tim::vx::ops::ReduceMean>, "Mean"),
+  REGISTER_OP_MAPPER(kTfLiteBuiltinLeakyRelu, LeakyReluMapper),
+  REGISTER_OP_MAPPER(kTfLiteBuiltinLogicalAnd, \
+      LogicalOpMapper<tim::vx::ops::LogicalAnd>, "And"),
+  REGISTER_OP_MAPPER(kTfLiteBuiltinLogicalOr, \
+      LogicalOpMapper<tim::vx::ops::LogicalOr>, "Or"),
+
```
Comment (sunshinemyson) [inline tensorflow/lite/delegates/vx-delegate/op_map.cc:1433]: blank line

**Review by sunshinemyson — CHANGES_REQUESTED** [review]

**Review by sunshinemyson — APPROVED** [review]

**Pull request merged**